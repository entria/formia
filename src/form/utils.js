import get from 'lodash.get';
import set from 'lodash.set';
import isEqual from 'lodash.isequal';

export const handleInitialValuesUpdate = (
  currentValues = {},
  currentInitialValues = {},
  newInitialValues = {},
) => {
  const updatedValues = { ...currentValues };

  Object.keys(newInitialValues).map(fieldName => {
    const fieldCurrentValue = get(currentValues, fieldName, null);
    const fieldCurrentInitialValue = get(currentInitialValues, fieldName, null);
    const isFieldPristine = isEqual(fieldCurrentValue, fieldCurrentInitialValue);

    if (!isFieldPristine) {
      return;
    }

    const fieldNewInitialValue = get(newInitialValues, fieldName, null);
    set(updatedValues, fieldName, fieldNewInitialValue);
  });

  return updatedValues;
};
