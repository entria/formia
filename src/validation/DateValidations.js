// @flow
import { isEmpty, Dates } from '@entria/utils';

import type { RuleError } from './Validation';

export const isValid = () => (value: any): ?RuleError => {
  if (isEmpty(value)) {
    return null;
  }

  if (Dates.isValid(value)) {
    return null;
  }

  return {
    code: 'Validation.Dates.isValid',
    message: 'Invalid date',
    params: [],
  };
};

export const isFuture = () => (value: any): ?RuleError => {
  if (isEmpty(value)) {
    return null;
  }

  if (Dates.isFuture(value)) {
    return null;
  }

  return {
    code: 'Validation.Dates.isFuture',
    message: 'Must be greater than today',
    params: [],
  };
};

export const isPast = () => (value: any): ?RuleError => {
  if (isEmpty(value)) {
    return null;
  }

  if (Dates.isPast(value)) {
    return null;
  }

  return {
    code: 'Validation.Dates.isPast',
    message: 'Must be lesser than today',
    params: [],
  };
};

export const isBetween = (minor: Date, major: Date) => (value: any): ?RuleError => {
  if (isEmpty(value)) {
    return null;
  }

  const isBiggerThanMinor = Dates.compare(value, minor) >= 0;
  const isLessThanMajor = Dates.compare(value, major) <= 0;
  if (isBiggerThanMinor && isLessThanMajor) {
    return null;
  }

  return {
    code: 'Validation.Dates.isBetween',
    message: `Must be greater than ${minor.toString()} and lesser than ${major.toString()}`,
    params: [minor, major],
  };
};
