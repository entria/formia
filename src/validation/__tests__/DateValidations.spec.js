import * as Dates from '../DateValidations';

import { setupTest } from '../../../test/utils';

beforeEach(async () => setupTest());

it('isValid - empty', () => {
  expect(Dates.isValid()()).toMatchSnapshot();
});

it('isValid - valid date', () => {
  expect(Dates.isValid()('2000-01-01')).toMatchSnapshot();
});

it('isValid - invalid date', () => {
  expect(Dates.isValid()('2000-01-35')).toMatchSnapshot();
});

it('isFuture - empty', () => {
  expect(Dates.isFuture()()).toMatchSnapshot();
});

it('isFuture - future date', () => {
  expect(Dates.isFuture()('2050-01-01')).toMatchSnapshot();
});

it('isFuture - past date', () => {
  expect(Dates.isFuture()('2000-01-01')).toMatchSnapshot();
});

it('isPast - empty', () => {
  expect(Dates.isPast()()).toMatchSnapshot();
});

it('isPast - past date', () => {
  expect(Dates.isPast()('2000-01-01')).toMatchSnapshot();
});

it('isPast - future date', () => {
  expect(Dates.isPast()('2050-01-01')).toMatchSnapshot();
});

it('isBetween - empty', () => {
  expect(Dates.isBetween('2000-01-01', '2000-12-31')()).toMatchSnapshot();
});

it('isBetween - is between', () => {
  expect(Dates.isBetween('2000-01-01', '2000-12-31')('2000-01-01')).toMatchSnapshot();
});

it('isBetween - is not between', () => {
  expect(Dates.isBetween('2000-01-01', '2000-12-31')('2050-01-01')).toMatchSnapshot();
});
