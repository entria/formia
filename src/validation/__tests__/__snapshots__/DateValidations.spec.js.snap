// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isBetween - empty 1`] = `null`;

exports[`isBetween - is between 1`] = `null`;

exports[`isBetween - is not between 1`] = `
Object {
  "code": "Validation.Dates.isBetween",
  "message": "Must be greater than 2000-01-01 and lesser than 2000-12-31",
  "params": Array [
    "2000-01-01",
    "2000-12-31",
  ],
}
`;

exports[`isFuture - empty 1`] = `null`;

exports[`isFuture - future date 1`] = `null`;

exports[`isFuture - past date 1`] = `
Object {
  "code": "Validation.Dates.isFuture",
  "message": "Must be greater than today",
  "params": Array [],
}
`;

exports[`isPast - empty 1`] = `null`;

exports[`isPast - future date 1`] = `
Object {
  "code": "Validation.Dates.isPast",
  "message": "Must be lesser than today",
  "params": Array [],
}
`;

exports[`isPast - past date 1`] = `null`;

exports[`isValid - empty 1`] = `null`;

exports[`isValid - invalid date 1`] = `
Object {
  "code": "Validation.Dates.isValid",
  "message": "Invalid date",
  "params": Array [],
}
`;

exports[`isValid - valid date 1`] = `null`;
